File: src/main/java/fr/ayfri/minecraft_art/gui/Slider.java
package fr.ayfri.minecraft_art.gui;

import fr.ayfri.minecraft_art.Main;
import processing.core.PApplet;
import processing.core.PConstants;

public class Slider {
	private final Main sketch;
	private final int x;
	private final int width;
	private final int height;
	private final float y;
	private float min = 1;
	private float max = 100;
	private int value;
	private float trueValue;
	public boolean isOn = false;
	
	public Slider(final Main sketch, final int x, final int y, final int width, final int height) {
		this.sketch = sketch;
		this.x = x;
		this.y = y;
		this.width = width;
		this.height = height;
		this.value = width / 2;
		this.trueValue = max / 2;
	}
	
	public void draw() {
		sketch.noStroke();
		sketch.fill(240);
		sketch.rect(x, y, width, height / 2f);
		sketch.fill(255);
		sketch.rect(x + value, y - height / 2f, 10, 3 * height / 2f);
		sketch.fill(0);
		sketch.textAlign(PConstants.CENTER, PConstants.CENTER);
		sketch.textSize(17f);
		sketch.text(String.valueOf(trueValue / 10f).substring(0, 3), x + width / 2f, y - 26);
	}
	
	public void update(final int value) {
		isOn = true;
		if (value - x < 0 || value - x > width) {
			return;
		}
		
		trueValue = PApplet.map(value - x, 0, width, min, max);
		if (trueValue < min) {
			trueValue = min;
		}
		if (trueValue > max) {
			trueValue = max;
		}
		
		this.value = value - x;
	}
	
	public float getValue() {
		return trueValue / 10f;
	}
	
	public boolean isOver() {
		return (x + width >= sketch.mouseX) && (sketch.mouseX >= x) && (y + height >= sketch.mouseY) && (sketch.mouseY >= y);
	}
	
	public void setMax(final float max) {
		this.max = max;
		this.trueValue = max / 2;
	}
	
	public void setMin(final float min) {
		this.min = min;
	}
}

File: src/main/kotlin/io/github/ayfri/minecraft_art/gui/Slider.kt
package io.github.ayfri.minecraft_art.gui

import io.github.ayfri.minecraft_art.Main
import processing.core.PApplet
import processing.core.PConstants

data class Slider(val sketch: Main, val x: Int, val y: Int, val width: Int, val height: Int) {
	private var handleOffset = width / 2
	private var trueValue = 50f

	var min = 1f
	var max = 100f
		set(value) {
			field = value
			trueValue = value / 2
		}

	/** Ratio applied to the input image, [trueValue] stays on a 0..[max] scale and is only scaled down when read. */
	val value get() = trueValue / 10f
	val isOver get() = sketch.mouseX in x..(x + width) && sketch.mouseY in y..(y + height)
	var isOn = false

	fun draw() = with(sketch) {
		noStroke()
		fill(240)
		rect(x.toFloat(), y.toFloat(), width.toFloat(), height / 2f)
		fill(255)
		rect((x + handleOffset).toFloat(), y - height / 2f, 10f, 3 * height / 2f)
		fill(0)
		textAlign(PConstants.CENTER, PConstants.CENTER)
		textSize(17f)
		text(value.toString().take(3), x + width / 2f, y - 26f)
	}

	fun update(mouseX: Int) {
		isOn = true
		val offset = mouseX - x
		if (offset < 0 || offset > width) return

		trueValue = PApplet.map(offset.toFloat(), 0f, width.toFloat(), min, max).coerceIn(min, max)
		handleOffset = offset
	}
}
